### PR TITLE
TSK-1943: added REST endpoints for requesting a review/changes on a Task

### PR DESCRIPTION
**Pull request by mustaphazorgati**
sonar: https://sonarcloud.io/summary/new_code?id=mustaphazorgati_taskana&branch=task%2F1943

---
Release Notes:

```
* TSK-1943: added the following REST endpoints:
 * /api/v1/tasks/{taskId}/request-review
 * /api/v1/tasks/{taskId}/request-review/force
 * /api/v1/tasks/{taskId}/request-changes
 * /api/v1/tasks/{taskId}/request-changes/force 
```
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] After integration of the PR, I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → TSK-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java [hunk 1]
```diff
@@ -202,6 +203,27 @@ public ResponseEntity<TaskRepresentationModel> claimTask(
     return ResponseEntity.ok(taskRepresentationModelAssembler.toModel(updatedTask));
   }
 
+  /**
+   * This endpoint requests a review on a Task.
+   *
+   * @param taskId the Id of the Task on which a review should be requested
+   * @return the review requested Task
```
Comment (ryzheboka) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:210]: Suggestion:
```
* This endpoint request a review on the specified Task
* 
* @param taskId the id of the relevant Task  
* @return the Task after a review has been requested
```

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 2]
```diff
@@ -540,7 +537,7 @@ void should_ReturnAllTasks_For_ProvidedSecondaryObjectReferenceByTypeAndValue()
     String url =
         restHelper.toUrl(RestEndpoints.URL_TASKS)
             + "?sor="
-            + URLEncoder.encode("{\"type\":\"Type2\",\"value\":\"Value2\"}", "UTF-8");
+            + URLEncoder.encode("{\"type\":\"Type2\",\"value\":\"Value2\"}", UTF_8);
```
Comment (ryzheboka) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:540]: Why did you change that?

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 3]
```diff
@@ -1395,6 +1392,62 @@ void should_ThrowException_When_CancelClaimingOfClaimedTaskByAnotherUser() {
         .isEqualTo(HttpStatus.BAD_REQUEST);
   }
 
+  @Test
+  void should_RequestReviewTask() {
```
Comment (ryzheboka) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1396]: What do you think about should_RequestReview() or should_RequestReviewOnATask() as the test name?

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 4]
```diff
@@ -1395,6 +1392,62 @@ void should_ThrowException_When_CancelClaimingOfClaimedTaskByAnotherUser() {
...
+
+    assertThat(cancelClaimResponse.getBody()).isNotNull();
+    assertThat(cancelClaimResponse.getStatusCode()).isEqualTo(HttpStatus.OK);
+    TaskRepresentationModel reviewTask = cancelClaimResponse.getBody();
+    assertThat(reviewTask.getOwner()).isNull();
+    assertThat(reviewTask.getState()).isEqualTo(TaskState.READY_FOR_REVIEW);
+  }
+
+  @Test
+  void should_RequestChangesTask() {
```
Comment (ryzheboka) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1424]: Same as for should_RequestReviewTask()?

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 5]
```diff
@@ -1395,6 +1392,62 @@ void should_ThrowException_When_CancelClaimingOfClaimedTaskByAnotherUser() {
...
+    assertThat(getTaskResponse.getBody()).isNotNull();
+    TaskRepresentationModel claimedTaskRepresentationModel = getTaskResponse.getBody();
+    assertThat(claimedTaskRepresentationModel.getState()).isEqualTo(TaskState.IN_REVIEW);
+    assertThat(claimedTaskRepresentationModel.getOwner()).isEqualTo("user-1-1");
+
+    // request changes
+    String url2 =
+        restHelper.toUrl(
+            RestEndpoints.URL_TASKS_ID_REQUEST_CHANGES, "TKI:000000000000000000000000000000000136");
+    ResponseEntity<TaskRepresentationModel> cancelClaimResponse =
```
Comment (ryzheboka) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: Please name the variable cancelClaimResponse according. E. g. requestChangesResponse

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 6]
```diff
@@ -1395,6 +1392,62 @@ void should_ThrowException_When_CancelClaimingOfClaimedTaskByAnotherUser() {
...
+    assertThat(getTaskResponse.getBody()).isNotNull();
+    TaskRepresentationModel claimedTaskRepresentationModel = getTaskResponse.getBody();
+    assertThat(claimedTaskRepresentationModel.getState()).isEqualTo(TaskState.CLAIMED);
+    assertThat(claimedTaskRepresentationModel.getOwner()).isEqualTo("user-1-1");
+
+    // request review
+    String url2 =
+        restHelper.toUrl(
+            RestEndpoints.URL_TASKS_ID_REQUEST_REVIEW, "TKI:000000000000000000000000000000000035");
+    ResponseEntity<TaskRepresentationModel> cancelClaimResponse =
```
Comment (ryzheboka) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: Please name the variable cancelClaimResponse according. E. g. requestReviewResponse

**Review by ryzheboka — CHANGES_REQUESTED** [review]

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 7]
```diff
@@ -1420,6 +1420,35 @@ void should_RequestReviewTask() {
     assertThat(repModel.getState()).isEqualTo(TaskState.READY_FOR_REVIEW);
   }
 
+  @Test
+  void should_ForceRequestReviewTask() {
```
Comment (ryzheboka) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: What do you think about should_ForceRequestReview_When_CurrentUserIsNotTheOwner?

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java (hunk 2 above)
Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:540]: IDE warnings :wink:

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java (hunk 1 above)
Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java]: done. Also done for forceRequestReview

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java (hunk 3 above)
Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1396]: should_RequestReviewOnATask() it is

Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: done

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 8]
```diff
@@ -1395,6 +1392,120 @@ void should_ThrowException_When_CancelClaimingOfClaimedTaskByAnotherUser() {
...
+
+    assertThat(requestedChangesResponse.getBody()).isNotNull();
+    assertThat(requestedChangesResponse.getStatusCode()).isEqualTo(HttpStatus.OK);
+    repModel = requestedChangesResponse.getBody();
+    assertThat(repModel.getOwner()).isNull();
+    assertThat(repModel.getState()).isEqualTo(TaskState.READY);
+  }
+
+  @Test
+  void should_ForceRequestChangesTask() {
```
Comment (ryzheboka) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1481]: What do you think about should_ForceRequestChanges_When_CurrentUserIsNotTheOwner?

**Review by ryzheboka — CHANGES_REQUESTED** [review]

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java (hunk 4 above)
Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: done

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java [hunk 9]
```diff
@@ -202,6 +203,69 @@ public ResponseEntity<TaskRepresentationModel> claimTask(
...
+          NotAuthorizedException {
+    Task task = taskService.requestReview(taskId);
+    return ResponseEntity.ok(taskRepresentationModelAssembler.toModel(task));
+  }
+
+  /**
+   * This endpoint requests changes on a Task.
+   *
+   * @param taskId the Id of the Task on which a review should be requested
+   * @return the change requested Task
```
[the remark this answers is not in the document]
Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java]: done

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java (hunk 7 above)
Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: done

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java (hunk 8 above)
Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: done

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java [hunk 10]
```diff
@@ -202,6 +203,90 @@ public ResponseEntity<TaskRepresentationModel> claimTask(
...
+    Task task = taskService.requestChanges(taskId);
+    return ResponseEntity.ok(taskRepresentationModelAssembler.toModel(task));
+  }
+
+  /**
+   * This endpoint force requests changes on a Task.
+   *
+   * @param taskId the Id of the Task on which a review should be requested
+   * @return the change requested Task
+   * @throws InvalidTaskStateException if the state of the Task with taskId is not IN_REVIEW
```
Comment (ryzheboka) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:274]: It doesn't hold because of the force option. How should we describe the Exception?

**Review by ryzheboka — APPROVED** [review]
